### ISSUE TITLE
We do not spot overridden pure virtual functions

### ISSUE THREAD

**Issue by adetaylor** [google/autocxx#491]
This is a problem encountered on an internal codebase, so I can't give details here, but recording that this bug exists for my own records. Once it's minimized I will post a test case here.

**Comment by adetaylor** [comment]
Closely related to #197.

The type is like this:

```cpp
class A {
public:
  virtual A* New() const override = 0;
};

class B : public A {
public:
  B* New() const override = 0;
};
```

We're not spotting the pure virtual function on B, thus we're generating `impl UniquePtr<B>` which generates invalid C++.

**Comment by martinboehme** [comment]
> The type is like this:
> 
> ```c++
> class A {
> public:
>   virtual A* New() const override = 0;

I think this should drop the `override`?

I find it a bit puzzling that this doesn't work today. Why would `B` look different to the analysis code than `A`? Both `A` and `B` are abstract, and both `A::New()` and `B::New()` are pure virtual? The only thing that's different here is the inheritance...

**Comment by adetaylor** [comment]
Yes, that snippet isn't exactly correct - instead see the test case added in #505.

The problem is with derived classes with pure virtual methods that aren't marked virtual. The analysis of whether a type is abstract is local and doesn't consider base classes. This was a known limitation when I did the initial pure virtual spotting work, and should be fixable.

That said, the problem in the internal codebase turns out to be (possibly) unrelated.- #506 - I'll need to minimize that.

**Comment by adetaylor** [comment]
Fixed by #510, but only if the base classes are on the allowlist. Hmmm.... more thought required.
